### PR TITLE
VPD:DDIMM DDR5 support for eBMC based systems

### PR DESCRIPTION
**Pull request by nkantesh**

### CODE REVIEW

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
I can see many format issues here. Please fix that.

File: vpd-parser/memory_vpd_parser.cpp [hunk 1]
```diff
@@ -36,72 +42,306 @@ static constexpr auto JEDEC_SDRAMCAP_RESERVED = 6;
...
+
+static const std::unordered_map<uint8_t, uint8_t> diePerPkgMap = {
+		{0, 1}, {2, 2}, {3, 4}, {4, 8}, {5, 16}};
+
+static const std::unordered_map<uint8_t, uint8_t> sdramDnstyPerDieMap = {
+		{0, 0}, {1, 4}, {2, 8}, {3, 12}, {4, 16}, {5, 24}, {6, 32}, {7, 48}, {8, 64}};
+
 auto memoryVpdParser::getDimmSize(Binary::const_iterator iterator)
 {
-    size_t tmp = 0, dimmSize = 0;
```
Comment (GiridhariKrishnan) [inline vpd-parser/memory_vpd_parser.cpp:41]: need to keep DDr4 check too along with DDR5

**Review by GiridhariKrishnan — COMMENTED** [review]
May need DDr4 related changes along with DDR5. Thank you

**Comment by nkantesh** [comment]
> I can see many format issues here. Please fix that.

done, please check now

File: vpd-parser/memory_vpd_parser.cpp (hunk 1 above)
Comment (nkantesh) [reply vpd-parser/memory_vpd_parser.cpp:41]: > need to keep DDr4 check too along with DDR5

DDR4 check is there, I have not removed, due to formatting issue you might have not found it. Please check now

File: const.hpp [hunk 2]
```diff
@@ -167,6 +167,18 @@ enum PelSeverity
     ERROR
 };
 
+enum SdRamType
+{
+	EN_DDR4 = 0x0C,
```
[the remark this answers is not in the document]
Comment (nkantesh) [reply const.hpp]: > Alignment may be necessary here.

Done

File: ibm_vpd_utils.cpp [hunk 3]
```diff
@@ -1023,5 +1023,20 @@ Binary getVpdDataInVector(const nlohmann::json& js, const std::string& file)
 
     return vpdVector;
 }
+void getNumOfEvenAndOddNumbrsInGivenNum(int& num, uint8_t& numOfEvenNumbrs, uint8_t& numOfOddNumbrs)
+{
+	uint8_t evenNumCount = 0;
+	uint8_t oddNumCount = 0;
+	while (num >= 0)
```
[the remark this answers is not in the document]
Comment (nkantesh) [reply ibm_vpd_utils.cpp]: > Can have normal check too reduces complexity.

Sorry, I didn't get you. we can discuss

File: const.hpp [hunk 4]
```diff
@@ -167,6 +167,18 @@ enum PelSeverity
     ERROR
 };
 
+enum SdRamType
+{
+    EN_DDR4 = 0x0C,
+    EN_DDR5 = 0x12
+};
+enum SdRamMixType
+
```
Comment (SunnySrivastava1984) [inline const.hpp:176]: Why this blank line here?

File: const.hpp [hunk 5]
```diff
@@ -167,6 +167,18 @@ enum PelSeverity
     ERROR
 };
 
+enum SdRamType
+{
+    EN_DDR4 = 0x0C,
+    EN_DDR5 = 0x12
+};
+enum SdRamMixType
```
Comment (SunnySrivastava1984) [inline const.hpp:175]: Please have a blank line between two block of codes.

File: ibm_vpd_utils.cpp [hunk 6]
```diff
@@ -1023,5 +1023,21 @@ Binary getVpdDataInVector(const nlohmann::json& js, const std::string& file)
 
     return vpdVector;
 }
+void getNumOfEvenAndOddNumbrsInGivenNum(int& num, uint8_t& numOfEvenNumbrs,
```
Comment (SunnySrivastava1984) [inline ibm_vpd_utils.cpp:1026]: Same here, a balnk line.

File: vpd-parser/memory_vpd_parser.hpp [hunk 7]
```diff
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "ibm_vpd_utils.hpp"
```
Comment (SunnySrivastava1984) [inline vpd-parser/memory_vpd_parser.hpp:3]: Why you have to add this?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: vpd-parser/memory_vpd_parser.hpp (hunk 7 above)
Comment (SunnySrivastava1984) [reply vpd-parser/memory_vpd_parser.hpp:3]: When you need in Cpp file add there not in header.

File: const.hpp (hunk 4 above)
Comment (nkantesh) [reply const.hpp:175]: > Why this blank line here?

clang-format was added, don't know why. I removed it

File: ibm_vpd_utils.cpp (hunk 6 above)
Comment (SunnySrivastava1984) [reply ibm_vpd_utils.cpp:1027]: I was not pointing that, I was asking to have a blank line between between two functions, which you have fixed. Thanks.

File: ibm_vpd_utils.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (GiridhariKrishnan) [reply ibm_vpd_utils.cpp]: It's taking O(n) can be reduced to O(1). As discussed can you please try to accommodate?

**Review by jinuthomas — CHANGES_REQUESTED** [review]
I would not like the current code being touched , you need to have a new function for the DDR5 size calculation and call it accordingly from the caller function. I would expect a simplified calculation for the even odd. and will ask to remove file specific constants and functions from the common files.

File: vpd-parser/memory_vpd_parser.cpp [hunk 8]
```diff
@@ -36,6 +42,53 @@ static constexpr auto JEDEC_SDRAMCAP_RESERVED = 6;
...
+    EN_DDR5 = 0x12
+};
+
+enum SdRamMixType
+{
+    EN_SYMMETRICAL = 0,
+    EN_ASYMMETRICAL = 1
+};
+
+static void getNumOfEvenAndOddNumbrsInGivenNum(int& num,
```
Comment (PriyangaRamasamy) [inline vpd-parser/memory_vpd_parser.cpp:88]: why did you declare this api getNumOfEvenAndOddNumbrsInGivenNum here and defined it at the end of file?
Why can't you declare and define here (line88) itself?

**Pull request closed without merging**